poc/vdaf_poplar1.sage

```diff
@@ -20,184 +21,183 @@ class Poplar1(Vdaf):
         # compute the sketch during preparation. This sketch is used
         # to verify the one-hotness of their output shares.
+
+        auth = prg.next_vec(Poplar1.Idpf.FieldInner,
+                                      Poplar1.Idpf.BITS - 1)
         beta_inner = [
-            [Poplar1.Idpf.FieldInner(1), k] \
-                for k in prg.next_vec(Poplar1.Idpf.FieldInner,
-                                      Poplar1.Idpf.BITS - 1) ]
-        beta_leaf = [Poplar1.Idpf.FieldLeaf(1)] + \
-            prg.next_vec(Poplar1.Idpf.FieldLeaf, 1)
+            [Poplar1.Idpf.FieldInner(1), k, Field2(1)]  \
+                for k in auth]
+        auth += prg.next_vec(Poplar1.Idpf.FieldLeaf, 1)
+        beta_leaf = [Poplar1.Idpf.FieldLeaf(1), auth[-1], Field2(1)]
 
         # Generate the IDPF keys.
         (public_share, keys) = \
             Poplar1.Idpf.gen(measurement, beta_inner, beta_leaf)
-
         # Generate correlated randomness used by the Aggregators to
         # compute a sketch over their output shares. PRG seeds are
         # used to encode shares of the `(a, b, c)` triples.
         # (See [BBCGGI21, Appendix C.4].)
-        corr_seed = [
-            gen_rand(Poplar1.Idpf.Prg.SEED_SIZE),
-            gen_rand(Poplar1.Idpf.Prg.SEED_SIZE),
-        ]
-        corr_prg = [
-            Poplar1.Idpf.Prg(corr_seed[0], dst + byte(0)),
-            Poplar1.Idpf.Prg(corr_seed[1], dst + byte(1)),
-        ]
-
-        # For each level of the IDPF tree, shares of the `(A, B)`
-        # pairs are computed from the corresponding `(a, b, c)`
-        # triple and authenticator value `k`.
-        corr_inner = [[], []]
+
+        # Generate the correction words
+        # For each level of the IDPF tree, the correction word
+        # is derived from the prefix of the measurement and the
+        # IDPF output shares of both aggregators.
+        corr_inner = []
         for level in range(Poplar1.Idpf.BITS):
+            prefix = (measurement // (2 ** (Poplar1.Idpf.BITS - 1 - level)))
             Field = Poplar1.Idpf.current_field(level)
-            k = beta_inner[level][1] if level < Poplar1.Idpf.BITS - 1 \
-                else beta_leaf[1]
-            (a, b, c) = vec_add(corr_prg[0].next_vec(Field, 3),
-                                corr_prg[1].next_vec(Field, 3))
-            A = -Field(2) * a + k
-            B = a^2 + b - a * k + c
-            corr1 = prg.next_vec(Field, 2)
-            corr0 = vec_sub([A, B], corr1)
-            if level < Poplar1.Idpf.BITS - 1:
-                corr_inner[0] += corr0
-                corr_inner[1] += corr1
+            k = auth[level]
+            print('randomness\t:', k)
+            (leader_data_share, leader_auth_share, leader_indicator_share) = \
+                Poplar1.Idpf.eval(0, public_share, keys[0], level, [prefix])[0]
+            helper_auth_share = leader_auth_share -k
+            helper_data_share = leader_data_share - Field(1)
+            w = Poplar1.hash1(level, prefix, \
```

Comment:
Isn't the correction word `w` supposed to be the XOR of the hashes, not the concatenation?

Reply:
It's not the XOR or the concatenation, it's the sum (we're working with additive shares not boolean hence the change from Mike's writeup)